Нестеров И.С. Б9123-01.03.02сп

Тема 1. Интерполирование функции с помощью
многочленов Лагранжа и многочленов Ньютона с разделенными разностями

In [11]:
from math import log, cos
from typing import Callable

1. На отрезке [a,b] получить таблицу значений функции y=f(x) в равноотстоящих точках $xi = a + ih;i = 0, 1, 2, . . . , 10; h = (b − a)/10$. Варианты
функции y = f(x) и отрезка [a,b], в моем случае отрезок [0.4, 0.9]

In [21]:
function = lambda x: x**2 + log(x)
abs_error = lambda x, y: abs(x - y) if x else "inf"
h = (0.9 - 0.4)/10
grid = [round(0.4 + h * i, 2) for i in range(11)]
print(*grid, sep='\n')

0.4
0.45
0.5
0.55
0.6
0.65
0.7
0.75
0.8
0.85
0.9


 2. С помощью интерполяционных формул Лагранжа и Ньютона с разделенными разностями выполнить линейную интерполяцию в точке $x^∗$. Допустима ли линейная интерполяция таблично заданной функции в точке $x^* (x_i<x∗<x_{i+1})$, обеспечивающая погрешность, не превосходящую $10^{−4}$?
Сравнить результаты со значением, получаемым при непосредственном вычислении по формуле $y^∗ = f(x^∗)$. 

In [15]:
def find_nearest(grid: list[float],
                 node: float = 0) -> int:
    """
    Эта функция возвращает индекс ближайшего по значению элемента из списка grid к числу node
    Args:
        grid (Optional[list[float]], optional): Список узлов сетки. Defaults to None.
        node (float, optional): Точка, к которой нужно найти ближайюшую из сетки. Defaults to 0.

    Returns:
        int: Индекс ближайшего элемента
    """
    l = 0
    r = len(grid) - 1
    while r - l > 1:
        avr = (r + l) // 2
        if grid[avr] > node:
            r = avr
        else:
            l = avr
    return l if abs(node - grid[l]) < abs(node - grid[r]) else r

$ L(x) = \sum_{i=0}^{n} f(x_i) * \prod_{j=0, j \neq i}^{n} \frac{x - x_j}{x_i - x_j} $

In [16]:

def lagrange_interpolate(function: Callable[[float], float], 
                         node: float = 0, grid: list[float] = None) -> float:
    """
        Функция, которая вычисляет интерполяционный многочлен Лагранжа
    Args:
        function (Callable[[float], float]): Функция, которую нужно интерполировать 
        node (float, optional): Узел. Defaults to 0.
        degree (int, optional): Степень полинома Лагранжа (). Defaults to 10.
        grid (Optional[list[float]], optional): Сетка узлов. Defaults to None.

    Returns:
        Optional[float]: Значения функции
    """
    interpolation_polynom = 0        
    for i in range(len(grid)):
        prod = 1
        
        for j in range(len(grid)):
            if i == j: continue
            prod *= (node - grid[j])/(grid[i] - grid[j])        
        interpolation_polynom += prod * function(grid[i])
        
    return interpolation_polynom

In [23]:
x_test = [0.52, 0.42, 0.87, 0.67]
for x in x_test:
    i = find_nearest(grid=grid, node=x)
    if grid[i] >= x:
        lin_grid = [grid[i-1 if i > 0 else 0], grid[i if i > 0 else 1]]    
        sqr_grid = [grid[i-1 if 0 < i < 9 else 0 if i == 0 else 7], grid[i if 0 < i < 9 else 1 if i == 0 else 8], grid[i+1 if 0 < i < 9 else 9 if i == 9 else 2]]
    else:
        lin_grid = [grid[i if i < 9 else 8], grid[i+1 if i < 9 else 9]]    
        sqr_grid = [grid[i-1 if 0 < i < 9 else 0 if i == 0 else 7], grid[i if 0 < i < 9 else 1 if i == 0 else 8], grid[i+1 if 0 < i < 9 else 9 if i == 9 else 2]]
    print(f'Линейная интерполяция: x = {x} ', function(x), lagrange_interpolate(function=function, node=x, grid=lin_grid))
    print('Ошибка интерполяции:', abs_error(function(x), lagrange_interpolate(function=function, node=x, grid=lin_grid)))
    print(f'Квадратичная интерполяция: x = {x}', function(x), lagrange_interpolate(function=function, node=x, grid=sqr_grid))
    print('Ошибка интерполяции:', abs_error(function(x), lagrange_interpolate(function=function, node=x, grid=sqr_grid)))


Линейная интерполяция: x = 0.52  -0.3835264674066639 -0.3840231086382153
Ошибка интерполяции: 0.000496641231551409
Квадратичная интерполяция: x = 0.52 -0.3835264674066639 -0.38341706833579514
Ошибка интерполяции: 0.00010939907086876532
Линейная интерполяция: x = 0.42  -0.6911005677047231 -0.6921775176116017
Ошибка интерполяции: 0.001076949906878677
Квадратичная интерполяция: x = 0.42 -0.6911005677047231 -0.6912868152117749
Ошибка интерполяции: 0.0001862475070518732
Линейная интерполяция: x = 0.87  0.6176379326664924 0.617230919228799
Ошибка интерполяции: 0.00040701343769333764
Квадратичная интерполяция: x = 0.87 0.6176379326664924 0.6175350274188807
Ошибка интерполяции: 0.0001029052476116954
Линейная интерполяция: x = 0.67  0.04842243340287483 0.04836027276903461
Ошибка интерполяции: 6.216063384022197e-05
Квадратичная интерполяция: x = 0.67 0.04842243340287483 0.04847244103141235
Ошибка интерполяции: 5.000762853751839e-05


$ P_n(x) = f(x_0) + \sum_{k=1}^{n}(f(x_0,...,x_k ) * \prod_{i=0}^{k-1}(x - x_i)) $

$ f(x_0,...,x_k) = \sum_{i=0}^{k}(\frac{f(x_i)}{\prod_{j=0, j \neq i}^{k}(x_i-x_j)}) $

In [18]:
def newton_interpolate(function: Callable[[float], float],
                       node: float,
                       grid: list[float]) -> float:
    """
    Эта функция высчитывает и возвращает интерполяционный полином Ньютона
    Args:
        function (Callable[[float], float]): Функция, которую нужно интерполировать
        node (float): Точка, в которой хотим найти значение
        grid (List[float]): Сетка узлов

    Returns:
        float: Значение интерполяционного полинома Ньютона в точке node
    """
    n = len(grid)    
    f_values = [function(x) for x in grid]
    
    divided_differences = [[0] * n for _ in range(n)]
    
    for i in range(n):
        divided_differences[i][0] = f_values[i]
    
    for j in range(1, n):
        for i in range(n - j):
            divided_differences[i][j] = (
                divided_differences[i + 1][j - 1] - divided_differences[i][j - 1]
            ) / (grid[i + j] - grid[i])
    
    interpolation_polynomial = divided_differences[0][0]
    product_term = 1.0
    
    for k in range(1, n):
        product_term *= (node - grid[k - 1])
        interpolation_polynomial += divided_differences[0][k] * product_term
    
    return interpolation_polynomial

In [ ]:
x_test = [0.52, 0.42, 0.87, 0.67]
for x in x_test:
    i = find_nearest(grid=grid, node=x)
    if grid[i] >= x:
        lin_grid = [grid[i-1 if i > 0 else 0], grid[i if i > 0 else 1]]    
        sqr_grid = [grid[i-1 if 0 < i < 9 else 0 if i == 0 else 7], grid[i if 0 < i < 9 else 1 if i == 0 else 8], grid[i+1 if 0 < i < 9 else 9 if i == 9 else 2]]
    else:
        lin_grid = [grid[i if i < 9 else 8], grid[i+1 if i < 9 else 9]]    
        sqr_grid = [grid[i-1 if 0 < i < 9 else 0 if i == 0 else 7], grid[i if 0 < i < 9 else 1 if i == 0 else 8], grid[i+1 if 0 < i < 9 else 9 if i == 9 else 2]]
    print(f'Линейная интерполяция: x = {x} ', function(x), newton_interpolate(function=function, node=x, grid=lin_grid))
    print('Ошибка интерполяции:', abs_error(function(x), newton_interpolate(function=function, node=x, grid=lin_grid)))
    print(f'Квадратичная интерполяция: x = {x}', function(x), newton_interpolate(function=function, node=x, grid=sqr_grid))
    print('Ошибка интерполяции:', abs_error(function(x), newton_interpolate(function=function, node=x, grid=sqr_grid)))

Линейная интерполяция: x = 0.52  -0.3835264674066639 -0.3840231086382153
Ошибка интерполяции: 0.000496641231551409
Квадратичная интерполяция: x = 0.52 -0.3835264674066639 -0.38341706833579514
Ошибка интерполяции: 0.00010939907086876532


С помощью формулы остаточного члена интерполяционной формулы Лагранжа первого порядка $R_1(x) = f^{′′}(ξ)ω_2(x)/2$, где $ξ ∈ [x_i, x_{i+1}]$, а
$ω_2(x) = (x−x_i)·(x−x_i+1)$, на отрезке $[x_i, x_i+1]$ оценить минимальное и максимальное значения $f^{′′}(x)$, а затем минимальное и максимальное значения остаточного члена $R_1(x)$.

Так как я выбрал функцию $ x^2 + ln(x) $ то её вторая производная будет иметь вид $ 2 - \frac{1}{x^2} $, тогда ее максимальное значения на отрезке $ [x_i, x_{i+1}] $ будет в конце отрезка, т.е. в точке $ x_{i+1} $, а минимум в точке $ x_{i} $ соответственно
Тогда, при $ x^* = 0.52 $ будем рассматривать отрезок $[0.5, 0.55]$. На этом отрезке $ min(f^{''}(\xi)) = f^{''}(0.5) = -2$, а $ max(f^{''}(\xi)) = f^{''}(0.55) = -1.3057851239669418 $ 

$ \omega_2(x) = (x - x_i)(x - x_{i+1}) = (0.52 - 0.5)(0.52 - 0.55) = -0.0003 $ 

$ max(R_1(0.52)) = -1.3057851239669418 * -0.00015 =  0.00019586776859504125$

$ min(R_1(0.52)) = -2 * -0.00015 = 0.0003 $

Проверим $0.00019586776859504125 < R_1(0.52) < 0.0003$

$ L1(x^∗) = f(x_i) · \frac{x^∗ − x_{i+1}}{x_i − x_{i+1}} + f(x_{i+1}) · \frac{x^∗ − x_i}{x_{i+1} − x_i} = -0.4431471805599453 * \frac{0.52 - 0.55}{0.5 - 0.55} - 0.29533700075562036 * \frac{0.52 - 0.5}{0.55 - 0.5}  = -0.3840231086382153 $

$ R_1(x^*) = L_1(x^*) - f(x^*) = L_1(0.52) - f(0.52) = -0.3840231086382153 + 0.3835264674066639 = -0.000496641231551409$

Неравенство не выполняется => погрешность может превосходить $ 10^{-4} $

Ответим на вопрос о том,  допустима ли квадратичная интерполяция таблично заданной функции в точке $x^∗$, обеспечивающая погрешность, не превосходящую $10^{−5}$?

По интерполяционной формуле Лагранжа второго порядка вычислим $L_2$

$ L_2(x) = f(x_{i-1}) \frac{(x^* - x_i)(x^* - x_{i+1})}{(x_{i-1} - x_i)(x_{i-1} - x_{i+1})} + f(x_i) \frac{(x^* - x_{i-1})(x^*-x_{i+1})}{(x_i - x_{i-1})(x_i - x_{i+1})} + f(x_{i+1}) \frac{(x^* - x_{i-1})(x^*-x_i)}{(x_{i+1}-x_{i-1})(x_{i+1}-x_i)}$, здесь $ x_{i-1} = 0.45, x_i=0.5, x_{i+1} = 0.55, x^* = 0.52 $

$ L_2(x) = -0.38341706833579514 $

$ R_2(x) = f^{′′′}(ξ)ω_3(x)/6, ξ ∈ [x_{i−1}, x_{i+1}], ω_3(x) = (x−x_{i−1})(x−x_i)(x−x_{i+1}) $

$ f^{′′′}(x) = \frac{2}{x^3} $, чем меньше х, тем больше значение производной => максимум будет достигаться $ x_{i-1} $, а минимум в $ x_{i+1} $

$ min(f^{′′′}(x)) = f^{′′′}(0.55) = 12.02103681442524, max(f^{′′′}(x)) = f^{′′′}(0.45) = 21.94787379972565 $

$ \omega_3(0.52) = -0.000042 $

$ min(R_2(x)) = 12.02103681442524 * -0.000042 / 6 = -0.000084 $

$ max(R_2(x)) = 21.94787379972565 * -0.000042 / 6 = -0.000154 $

$ R_2(0.52) = -0.38341706833579514 + 0.3835264674066639 = 0.00010939907086876532 $

Насколько я понимаю, тут должен быть модуль, но в методичке его нет... Если без него, то неравенство не выполняется, если с ним, то выполняется.
Я думаю, что модуль должен быть
